### PR TITLE
Initial vfio plugin

### PR DESCRIPTION
**Pull request by r-rojo**
Begin implementing a plugin for vfio.
* Rename plugin directory to plugins
* Add plugins/vfio/plugin.c
  * This implements the plugin interfaces
    * opae_plugin_configure
      * Fills in adapter table for functions implemented
    * opae_plugin_initialize
        * calls pci_discover, features_discover
    * opae_plugin_finalize
        * calls free_tokens_list, free_device_list
* Add plugins/vfio/opae_vfio.c
  * Implment pci_discover
      * Scans /sys/bus/pci/drivers/vfio-pci/* to look for pci devices bound
  * Implment features_discover
      * For every pci device discovered, look in its mmio space for OPAE objects (FPGA_DEVICE, FPGA_ACCELERATOR)
      * First look in BAR0 for known FME GUIDs
        * If found, look for FME/Port features using legacy layout
      * If no known FME GUID is found, treat every available BAR as an  FPGA_ACCELERATOR object
        This allows user space client code to implement any user mode driver(s) using OPAE APIs
  * Implement cleanup routines called in opae_plugin_finalize
 * Add plugins/vfio/dfl.c
  * Implement routines for walking BARs similar to dfl driver

### CODE REVIEW

**Review by anandaravuri — DISMISSED** [review]
Good work

File: plugins/vfio/opae_vfio.c [hunk 1]
```diff
@@ -0,0 +1,919 @@
...
+	return FPGA_OK;
+
+}
+
+fpga_result vfio_fpgaCloneToken(fpga_token src, fpga_token *dst)
+{
+	vfio_token *_src = (vfio_token*)src;
+	vfio_token *_dst;
+	if (!src || !dst) {
+		printf("src or dst is NULL\n");
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:801]: use ERR( )

File: plugins/vfio/opae_vfio.c [hunk 2]
```diff
@@ -0,0 +1,919 @@
...
+		return FPGA_INVALID_PARAM;
+	}
+	if (_src->magic != VFIO_TOKEN_MAGIC) {
+		printf("Invalid src\n");
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:805]: use ERR( )

File: plugins/vfio/opae_vfio.c [hunk 3]
```diff
@@ -0,0 +1,919 @@
...
+
+	(void)flags;
+	vfio_handle *h = (vfio_handle*)handle;
+	struct opae_vfio *v = (struct opae_vfio*)&h->vfio_device;
+	uint8_t *virt = NULL;
+	uint64_t iova = 0;
+	size_t sz = len > HUGE_2M ? ROUND_UP(len, HUGE_1G) :
+		    len > 4096 ? ROUND_UP(len, HUGE_2M) : 4096;
+	if (opae_vfio_buffer_allocate(v, &sz, &virt, &iova)) {
+		printf("could not allocate buffer\n");
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:848]: use ERR( )

File: plugins/vfio/opae_vfio.c [hunk 4]
```diff
@@ -0,0 +1,919 @@
...
+		return FPGA_EXCEPTION;
+	}
+	vfio_buffer *buffer = (vfio_buffer*)malloc(sizeof(vfio_buffer));
+	if (!buffer) {
+		printf("error allocating buffer metadata\n");
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:853]: use ERR( )

File: plugins/vfio/opae_vfio.c [hunk 5]
```diff
@@ -0,0 +1,919 @@
...
+		if (opae_vfio_buffer_free(v, virt)) {
+			printf("error freeing vfio buffer\n");
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:855]: use ERR( )

File: plugins/vfio/opae_vfio.c [hunk 6]
```diff
@@ -0,0 +1,919 @@
...
+		}
+		return FPGA_NO_MEMORY;
+	}
+	memset(buffer, 0, sizeof(vfio_buffer));
+
+	if (_vfio_buffers) {
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:861]: Need a lock to protect _vfio_buffers.

File: plugins/vfio/opae_vfio.c [hunk 7]
```diff
@@ -0,0 +1,919 @@
...
+
+	vfio_handle *h = (vfio_handle*)handle;
+	struct opae_vfio *v = (struct opae_vfio*)&h->vfio_device;
+
+	vfio_buffer *ptr = _vfio_buffers;
+	vfio_buffer *prev = NULL;
+	while (ptr) {
+		if (ptr->wsid == wsid) {
+			if (opae_vfio_buffer_free(v, ptr->virtual)) {
+				printf("error freeing vfio buffer\n");
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:887]: use ERR( )

File: plugins/vfio/opae_vfio.c [hunk 8]
```diff
@@ -0,0 +1,919 @@
...
+			}
+			if (prev) {
+				prev->next = ptr->next;
+			} else {
+				_vfio_buffers = ptr->next;
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:892]: Need a lock to protect _vfio_buffers.

File: plugins/vfio/opae_vfio.c [hunk 9]
```diff
@@ -0,0 +1,919 @@
...
+	}
+
+	return FPGA_NOT_FOUND;
+}
+
+fpga_result vfio_fpgaGetIOAddress(fpga_handle handle, uint64_t wsid,
+			          uint64_t *ioaddr)
+{
+	(void)handle;
+	vfio_buffer *ptr = _vfio_buffers;
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.c:908]: Need to protect _vfio_buffers with a lock.

File: plugins/vfio/opae_vfio.h [hunk 10]
```diff
@@ -0,0 +1,73 @@
+#ifndef _OPAE_VFIO_PLUGIN_H
```
Comment (tswhison) [inline plugins/vfio/opae_vfio.h:1]: Add (c) header.

File: plugins/vfio/plugin.c [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+#include <stdlib.h>
+#include <dlfcn.h>
+
+#include <opae/types_enum.h>
+
+#include "adapter.h"
+#include "opae_vfio.h"
+
+
+int __XFPGA_API__ vfio_plugin_initialize(void)
```
Comment (tswhison) [inline plugins/vfio/plugin.c:39]: Should make a new #define name for your plugin.  __VFIO_API__?

File: plugins/vfio/plugin.c [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+{
+	int res = pci_discover();
+	if (res) {
+		printf("error with pci_discover\n");
```
Comment (tswhison) [inline plugins/vfio/plugin.c:43]: use ERR( )

File: plugins/vfio/plugin.c [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	res = features_discover();
+	if (res) {
+		printf("error discovering features\n");
```
Comment (tswhison) [inline plugins/vfio/plugin.c:47]: use ERR( )

File: plugins/vfio/dfl.c [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include "opae_vfio.h"
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply plugins/vfio/dfl.c:26]: 👍

File: plugins/vfio/opae_vfio.c (hunk 8 above)
Comment (tswhison) [reply plugins/vfio/opae_vfio.c:933]: Lock for _vfio_buffers?

File: plugins/vfio/opae_vfio.c (hunk 9 above)
Comment (tswhison) [reply plugins/vfio/opae_vfio.c:948]: Lock for _vfio_buffers?

File: plugins/vfio/opae_vfio.c [hunk 15]
```diff
@@ -0,0 +1,919 @@
...
+	res = FPGA_OK;
+
+out:
+	regfree(&re);
+	return res;
+}
+
+void free_device_list()
+{
+	while (_pci_devices) {
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply plugins/vfio/opae_vfio.c:154]: It's filled in at initialization and isn't updated at runtime right now

File: plugins/vfio/opae_vfio.c [hunk 16]
```diff
@@ -0,0 +1,919 @@
...
+		pci_device_t *p = _pci_devices;
+		_pci_devices = _pci_devices->next;
+		free(p);
+	}
+}
+
+void free_token_list()
+{
+	while (_vfio_tokens) {
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply plugins/vfio/opae_vfio.c:163]: It's filled in at initialization and isn't updated at runtime right now

File: plugins/vfio/opae_vfio.c [hunk 17]
```diff
@@ -0,0 +1,959 @@
...
+}
+
+static int read_pci_attr_u32(const char *addr, const char *attr, uint32_t *value)
+{
+	char str_value[64];
+	char *endptr = NULL;
+	int res = read_pci_attr(addr, attr, str_value, sizeof(str_value));
+	if (res) return res;
+	uint32_t v = strtoul(str_value, &endptr, 0);
+	if (!v && str_value == endptr) {
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply plugins/vfio/opae_vfio.c:117]: if str_value has non-numeric characters after the numeric ones, endptr will point to that.
Take this example:
```C++
// Example program
#include <iostream>
#include <string>

int main()
{
    for (auto s : { "abc12abc", "12abc", "abc12", "12" }) {
        char *endptr;
        uint32_t v = strtoul(s, &endptr, 0);
        std::cout << "string: " << s << ", value: " << v << ", endptr: " << endptr << "\n";
        v = strtoul("abc", &endptr, 0);

    }
}
```

The output of this is:
```terminal
string: abc12abc, value: 0, endptr: abc12abc
string: 12abc, value: 12, endptr: abc
string: abc12, value: 0, endptr: abc12
string: 12, value: 12, endptr:
```

You can try this example here (if still available):
http://cpp.sh/8wv73

[the remark this answers is not in the document]
Comment (tswhison) [reply plugins/vfio/opae_vfio.c:117]: Right, so in other words, if endptr points to the end of your input string after the call, then the conversion was successful.  ie
```c
if (endptr != str_value + strlen(str_value))
```
If endptr is not equal to the string + the string length, then there was a conversion error.

**Review by tswhison — APPROVED** [review]
Awesome work!

**Review by asgardkm — APPROVED** [review]

**Pull request merged**